Arduino/MiP_ProMini_Pack_Library/src/mip.cpp

```diff
@@ -1937,6 +1951,251 @@ int8_t MiP::rawGetUserData(uint8_t address, uint8_t& userData)
+void MiP::rawSetMiPDetectionMode(uint8_t id, uint8_t txPower)
+{
+    uint8_t command[1+2];
+
+    MIP_ASSERT( MIP_IR_DETECTION_MODE_DISABLE <= id && id <= 0xFF );
+    MIP_ASSERT( 0x01 <= txPower && txPower <= 0x78 );
+
+    command[0] = MIP_CMD_SET_DETECTION_MODE;
+    command[1] = id;
+    command[2] = txPower;
+
+    rawSend(command, sizeof(command));
+}
+
+void MiP::enableIRRemoteControl()
+{
+    verifiedIRRemoteControl(MIP_IR_REMOTE_CONTROL_ENABLE);
+}
+
+void MiP::disableIRRemoteControl()
+{
+    verifiedIRRemoteControl(MIP_IR_REMOTE_CONTROL_DISABLE);
+}
+
+bool MiP::isIRRemoteControlEnabled()
+{
+    const uint8_t remoteControlEnabled[1] = { MIP_CMD_GET_IR_REMOTE_CONTROL };
+    uint8_t       response[1+1];
+    size_t        responseLength;
+    int8_t        result;
+
+    result = rawReceive(remoteControlEnabled, sizeof(remoteControlEnabled), response, sizeof(response), responseLength);
+
+    if (result)
+    {
+        return result;
+    }
+    if (responseLength != sizeof(response) ||
+        response[0] != MIP_CMD_GET_IR_REMOTE_CONTROL)
+    {
+        return MIP_ERROR_BAD_RESPONSE;
+    }
+
+    return response[1] == MIP_IR_REMOTE_CONTROL_ENABLE ? true : false;
+}
+
+void MiP::sendIRDongleCode(uint8_t sendCode[], uint8_t dataNumbers, uint8_t transmitPower)
+{
+    // According to WowWee documentation, indicate number of valid bits in transmission.
+    MIP_ASSERT( dataNumbers == 0x02 || dataNumbers == 0x03 || dataNumbers == 0x04 );
+
+    // Check for valid transmitPower values.  Must be between 1 and 120.
+    MIP_ASSERT( 0x01 <= transmitPower && transmitPower <= 0x78 );
+
+    // Check for valid codes in a three-byte send.
+    if(dataNumbers == 0x03 && sendCode[3] != 0xFF)
+        MIP_ASSERT( 0x00 <= sendCode[1] && sendCode[1] <= 0x03 );
+
+    // Check for valid codes in a four-byte send.
+    if(dataNumbers == 0x04 && sendCode[2] != 0xFF && sendCode[3] != 0xFF){
+        MIP_ASSERT( 0x00 <= sendCode[0] && sendCode[0] <= 0x03 );
+        MIP_ASSERT( 0x00 <= sendCode[1] && sendCode[1] <= 0x03 );
+    }
+    
+    uint8_t dataBits;
+    
+    // MiP's API requires the usable data be specified in bits lengths of 0x10, 0x18 or 0x20.
+    switch(dataNumbers)
+    {
+        case 0x02:
+            dataBits = 0x10;
+            break;
+        case 0x03:
+            dataBits = 0x18;
+            break;
+        case 0x04:
+            dataBits = 0x20;
+            break;
+        default:
+            dataBits = 0x00;
+            break;
+    }
+    
+    // Send this command blindly with no error checking since there is no easy way to determine if it has failed.
+    rawSendIRDongleCode(sendCode, dataBits, transmitPower);
+}
+
+// This internal protected method sends the send IR dongle code command with no error checking.
+// The error handling and recovery happens at a higher level of the driver.
+void MiP::rawSendIRDongleCode(uint8_t sendCode[], uint8_t dataNumbers, uint8_t transmitPower)
```

Comment:
Suggestion: Can probably just pull this raw function directly into the single function where it is called.